pkg/provisioner/ironic/ironic.go

```diff
@@ -508,23 +508,30 @@ func (p *ironicProvisioner) InspectHardware() (result provisioner.Result, detail
 				result.RequeueAfter = introspectionRequeueDelay
 				err = nil
 				return
+			case nodes.Manageable:
+				inspect = true
 			default:
-				p.log.Info("starting new hardware inspection")
-				var success bool
-				success, result, err = p.tryChangeNodeProvisionState(
-					ironicNode,
-					nodes.ProvisionStateOpts{Target: nodes.TargetInspect},
-				)
-				if success {
-					p.publisher("InspectionStarted", "Hardware inspection started")
-				}
-				return
+				inspect = true
 			}
+		} else {
+			err = errors.Wrap(err, "failed to extract hardware inspection status")
+			return
+		}
+	}
+	if reinstrospect && !(nodes.ProvisionState(ironicNode.ProvisionState) == nodes.Inspecting || nodes.ProvisionState(ironicNode.ProvisionState) == nodes.InspectWait) {
+		p.log.Info("starting new hardware inspection")
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(
+			ironicNode,
+			nodes.ProvisionStateOpts{Target: nodes.TargetInspect},
+		)
+		if success {
+			p.publisher("InspectionStarted", "Hardware inspection started")
 		}
-		err = errors.Wrap(err, "failed to extract hardware inspection status")
 		return
 	}
-	if !status.Finished {
+	if !status.Finished || (nodes.ProvisionState(ironicNode.ProvisionState) == nodes.Inspecting || nodes.ProvisionState(ironicNode.ProvisionState) == nodes.InspectWait) {
```

Comment:
In the past we haven't bothered to wait until the node changes state in Ironic if the data is available. Why do you think it is necessary?

Reply:
Pardon if I misunderstand the goal of the patch, but if we allow re-introspection, the data will be available from the beginning (the old data)

Reply:
It sounds like we should tell inspector to delete the old data first? Or is waiting for the state change good enough? If we aren't doing that now, it's likely a bug in the original implementation anyway.

Reply:
I don't think there is a call to remove old data (you could argue we should remove it when starting inspection.. maybe we should). Waiting for the state change will do the trick though.